src/external/dftk_to_wannier90.jl

```diff
@@ -0,0 +1,605 @@
+    @assert size(i_nn_kpts,1) == 2
+    char_to_vec(line) = [parse.(Int,x) for x in split(line,' ',keepempty=false)]
+    
+    nn_num = parse(Int64,ln[i_nn_kpts[1]+1])                                       #number of neighbours per k points
+    nn_kpts =  [ char_to_vec(l) for l in ln[i_nn_kpts[1]+2:i_nn_kpts[2]-1] ]       
+
+
+
+    ## Extract projections block : NEEDED FOR WIN GUESSES
+    i_projs = findall(x-> endswith(x,"projections"),ln)                                  #Indices of the first and last line of the projection block
+    @assert size(i_projs,1) == 2
+    raw_projs = [split(ln[i],' ',keepempty = false) for i in i_projs[1]+1:i_projs[2]-1]  #data in strings
+
+    # reshape so that one line gives all infos about one projection
+    n_projs = parse(Int,only(popfirst!(raw_projs)))
+    @assert(n_projs == size(raw_projs,1)/2)
+    raw_projs = reshape(raw_projs,(2,n_projs))
+
+    # Parse in the format proj = [ [[center],[l,mr,r],[z_axis],[x_axis], α], ... ]
+    projs = []
+    for j in 1:n_projs
+        center = [parse(Float64,x) for x in raw_projs[1,j][1:3]]
+        quantum_numbers = [parse(Int,x) for x in raw_projs[1,j][4:end] ]
+        z_axis = [parse(Float64,x) for x in raw_projs[2,j][1:3]]
+        x_axis = [parse(Float64,x) for x in raw_projs[2,j][4:6]]
+        α = parse(Float64,raw_projs[2,j][7])
+        push!(projs, [center,quantum_numbers,z_axis,x_axis,α])
+    end
+    
+    nn_kpts,nn_num,projs    
+end
+
+
+
+"""
+Take scf_res, the result of a self_consistent_field calculation and writes eigenvalues (in eV) in a format readable by Wannier90.
+"""
+function generate_eig_file(prefix::String,scf_res)
+    Ha_to_eV = 1/(unit_to_au(:eV))
+
+    eigvalues = scf_res.eigenvalues
+    k_size = size(eigvalues,1)
+    n_bands = size(eigvalues[1],1)
+
+    #write file
+    open("$prefix.eig","w") do f
+        for k in 1:k_size
+            for n in 1:n_bands
+                write(f,"  $n  $k  $(Ha_to_eV*eigvalues[k][n])"*"\n")
+            end
+        end
+    end
+
+    return nothing
+    
+end
+
+
+@doc raw"""
+ Computes the matrix ``[M^{k,b}]_{m,n} = \langle u_{m,k} | u_{n,k+b} \rangle`` for given k, kpb = k+b.
+ 
+ Return in the format of a (n_bands^2, 2) matrix, where each line is one overlap, and columns are real and imaginary parts.
+
+ K_shift is the "shifting" vector, correction due to the periodicity conditions imposed on k -> ψ_k. 
+ It is non zero if kpb is taken in another unit cell of the reciprocal lattice.
+ We use here that : ``u_{n(k + K_shift)}(r) = e^{-i*\langle K_shift,r \rangle} u_{nk}``
+"""
+function overlap_Mmn_k_kpb(pw_basis::PlaneWaveBasis,ψ,k,kpb,n_bands; K_shift = [0,0,0])
+    Mkb = zeros(Float64,n_bands*n_bands,2)
+    current_line = 0 #Manual count necessary to match the format of .mmn files.
+
+    for n in 1:n_bands
+        for m in 1:n_bands
+            
+            ovlp = 0im
```

Comment:
careful here, it's a type-instability. Try doing `ovlp = 0im; ovlp += randn(ComplexF64)`: ovlp changes type. One reason for bad performance (probably not the only one though)

Reply:
I see : ovlp = 0im gives a Complex{Int64}. As I understand it doing ovlp = zero(ComplexF64) should be better.

Reply:
Yup. Typically we try to never put floating point types explicitly but try to guess it from the environment (eg eltype(basis)). Although since you're going to write it to a file, might as well do F64 explicitly